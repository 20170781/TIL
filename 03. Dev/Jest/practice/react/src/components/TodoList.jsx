import React from 'react';
import TodoItem from './TodoItem';
import AddForm from './AddForm';

const TodoList = ({
  todos,
  onIncrement,
  onDecrement,
  onDelete,
  onAdd,
  onReset,
}) => {
  return (
    <>
      <AddForm onAdd={onAdd} />
      <ul className="todos">
        {todos.map((todo) => (
          <TodoItem
            key={todo.id}
            todo={todo}
            onIncrement={onIncrement}
            onDecrement={onDecrement}
            onDelete={onDelete}
          />
        ))}
      </ul>
      <button className="todos-reset" onClick={onReset}>
        Reset All
      </button>
    </>
  );
};

export default TodoList;
