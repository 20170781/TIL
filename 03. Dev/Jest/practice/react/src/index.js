import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './app/app';
import '@fortawesome/fontawesome-free/js/all.js';
import TodoPresenter from './logics/todo_presenter';

const todoPresenter = new TodoPresenter([
  { id: 1, name: 'JavaScript', count: 0 },
  { id: 2, name: 'TypeScript', count: 0 },
  { id: 3, name: 'React', count: 0 },
]);

ReactDOM.render(
  <React.StrictMode>
    <App presenter={todoPresenter} />
  </React.StrictMode>,
  document.getElementById('root')
);
