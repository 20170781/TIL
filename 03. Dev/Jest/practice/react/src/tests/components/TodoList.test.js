import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoList from '../../components/TodoList';

describe('TodoList', () => {
  const todos = [
    { name: 'javascript', count: 1, id: 1 },
    { name: 'typescript', count: 4, id: 2 },
  ];

  let TodoListComp;
  let onIncrement;
  let onDecrement;
  let onDelete;
  let onAdd;
  let onReset;

  beforeEach(() => {
    onIncrement = jest.fn();
    onDecrement = jest.fn();
    onDelete = jest.fn();
    onAdd = jest.fn();
    onReset = jest.fn();

    TodoListComp = (
      <TodoList
        todos={todos}
        onIncrement={onIncrement}
        onDecrement={onDecrement}
        onDelete={onDelete}
        onAdd={onAdd}
        onReset={onReset}
      />
    );
  });

  test('snapshot', () => {
    const component = render(TodoListComp);
    expect(component.container).toMatchSnapshot();
  });

  describe('Button click', () => {
    beforeEach(() => {
      render(TodoListComp);
    });

    test('calls onIncrement', () => {
      const button = screen.getAllByTitle('increase')[0];
      userEvent.click(button);
      expect(onIncrement).toHaveBeenCalledWith(todos[0]);
    });

    test('calls onDecrement', () => {
      const button = screen.getAllByTitle('decrease')[0];
      userEvent.click(button);
      expect(onDecrement).toHaveBeenCalledWith(todos[0]);
    });

    test('calls onDelete', () => {
      const button = screen.getAllByTitle('delete')[0];
      userEvent.click(button);
      expect(onDelete).toHaveBeenCalledWith(todos[0]);
    });

    test('calls onReset', () => {
      const button = screen.getByText('Reset All');
      userEvent.click(button);
      expect(onReset).toHaveBeenCalledTimes(1);
    });
  });
});
