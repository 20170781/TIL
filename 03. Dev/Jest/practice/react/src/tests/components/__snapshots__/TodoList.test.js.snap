// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList snapshot 1`] = `
<div>
  <form
    class="add-form"
  >
    <input
      class="add-input"
      placeholder="Todo"
      type="text"
    />
    <button
      class="add-button"
    >
      Add
    </button>
  </form>
  <ul
    class="todos"
  >
    <li
      class="todo"
    >
      <span
        class="todo-name"
        data-testid="todo-name"
      >
        javascript
      </span>
      <span
        class="todo-count"
        data-testid="todo-count"
      >
        1
      </span>
      <button
        class="todo-button todo-increase"
        title="increase"
      >
        <i
          class="fas fa-plus-square"
        />
      </button>
      <button
        class="todo-button todo-decrease"
        title="decrease"
      >
        <i
          class="fas fa-minus-square"
        />
      </button>
      <button
        class="todo-button todo-delete"
        title="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </li>
    <li
      class="todo"
    >
      <span
        class="todo-name"
        data-testid="todo-name"
      >
        typescript
      </span>
      <span
        class="todo-count"
        data-testid="todo-count"
      >
        4
      </span>
      <button
        class="todo-button todo-increase"
        title="increase"
      >
        <i
          class="fas fa-plus-square"
        />
      </button>
      <button
        class="todo-button todo-decrease"
        title="decrease"
      >
        <i
          class="fas fa-minus-square"
        />
      </button>
      <button
        class="todo-button todo-delete"
        title="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </li>
  </ul>
  <button
    class="todos-reset"
  >
    Reset All
  </button>
</div>
`;
